### ISSUE TITLE
Resolve getting the application log & trace issues.

### ISSUE THREAD

**Issue by ddebrunner** [IBMStreams/streamsx.topology#1302]
Python REST wrapper currently has a method `get_application_logs()` on the `Job` object: [1]
Its summary is:
_Retrieves the application log and trace files of the job and saves them as a tar file to the specified path with the given name._

The REST operation being wrapped is:
- an HTTP GET on `applicationLogTrace` (which will return a stream containing the compressed logs)
- a store of returned stream as a compressed tar file.

It takes a `path` to store the tar file and `prefix` of the file name, returning the file name.

It has been proposed to change this in two ways:

1. Change the name to `store_logs` or maybe `retrieve_application_logs`
   - With summary: _Download the console logs and application traces from a job and store them locally in a compressed 'tar' archive_
2. Simplify the api to just have an optional `filename` parameter that defines the filename to use for the tar file.

We need to resolve which direction we want to take.

[1]: https://github.com/IBMStreams/streamsx.topology/blob/d2daaaecee5f1953ee9e50977767609358e8e27e/com.ibm.streamsx.topology/opt/python/packages/streamsx/rest_primitives.py#L407

**Comment by ddebrunner** [comment]
I think it's important to note the call is getting the application *log* and *trace*. If in the future there were calls to just get the application log or trace individually, how would we evolve to that?

If it `store_logs` implies both, is it clear that `store_log` would just be *log*?

Or maybe a `log_only` parameter, though then we have to have a `log_only` and `trace_only parameter`, or maybe it's a signature of something like: `include_log=True, include_trace=True`.

**Comment by ddebrunner** [comment]
@wmarshall484 @ejpring  - Please comment.

**Comment by ejpring** [comment]
I propose:

**Job::retrieve_logs_traces(filename=None)**
Retrieve console logs and application traces from all of the job's PEs and store them as a compressed 'tar' archive in the specified file. If not specified, the default filename is 'jobname.tar.gz' in the current directory, where 'jobname' is the name assigned to the job when it was submitted. The method returns the name of the file it stored.

**PE::retrieve_console_log(filename=None)**
Retrieve the PE's console log and store it as uncompressed text in the specified file. If not specified, the default filename is 'PE_id.log', where 'id' is the PE's identifying number. The method returns the name of the file it stored.

**PE::retrieve_application_trace(filename=None)**
Retrieve the PE's application trace and store it as uncompressed text in the specified file. If not specified, the default filename is 'PE_id.trace', where 'id' is the PE's identifying number. The method returns the name of the file it stored.

**Comment by ddebrunner** [comment]
I would actually like the default file name to be unique, so that a monitoring script can be written that fetches log/trace periodically and not force the client to come up with a unique naming scheme.

With that, I also think that there exists the need to allow the directory to be specified when using the default name, the current directory is not always the best place.  This would then would then change the signatures to be (including being consistent with the names):
- `Job.retrieve_log_trace(dir=None, name=None)`
   - `name` defaults to ???
   - `jobname` does not work as a file name as it can include arbitrary characters, so what should the default be?
- `PE.retrieve_log(dir=None, name=None)`
    - name defaults to `PE_id_unique.log`
       - `id` is the PE identifier
       - `unique` is a value that makes the filename unique - from `tempfile`.
- `PE::retrieve_trace(dir=None, name=None)`
    - name defaults to `PE_id_unique.trace`
       - `id` is the PE identifier
       - `unique` is a value that makes the filename unique - from `tempfile`.

**Comment by ddebrunner** [comment]
Might be more natural to have `name=None, dir=None` ...

**Comment by ddebrunner** [comment]
Might be good to be consistent with `streamtool` ...

**Comment by ejpring** [comment]
Sorry, I really dislike the idea of creating collections of randomly named files. There is another way for a monitoring script to implement incremental log& trace retrieval that you might want to consider. The responses for all log and trace GET requests include the HTTP header 'Last-Modified-Milliseconds'. Its value can be cached and specified in the URL of the next GET request with the 'startTime' argument to retrieve only new log messages. These can then be appended to the file named by the caller.

**Comment by ddebrunner** [comment]
But that's forcing work on each user, for a common case. I think streamtool uses a "random" name or at least a unique name.

**Comment by ddebrunner** [comment]
Ok - `streamtool getapplicationlog` uses the strange fixed name `StreamsAppProfile.tgz` but does allow a `--colection-directory` option.

**Comment by ddebrunner** [comment]
So I think to retrieve only new messages a `since` parameter would be added, but no need to do that yet.

Maybe only in that case do we add a unique tag with it being the `since` argument.

**Comment by ddebrunner** [comment]
I was thinking of the console downloads, where a download of the logs for a job resulted in:

`StreamsJobLogs-standard1.3a4b0956-ec2c-4cd0-98e7-9da7c70c0fa3.job5-5656365070533058228.tgz`

I think that's `StreamsJobLogs-DOMAIN_ID.INSTANCE_ID.jobJOB_ID-5656365070533058228.tgz`

Not sure what the last element is.

**Comment by wmarshall484** [comment]
> Sorry, I really dislike the idea of creating collections of randomly named files.
I think the default of creating a unique filename is preferable in the absence of a given name. Creating a unique filename in which to store the logs removes the possibility of a user accidentally modifying or overwriting previously downloaded logs. If the user wishes to append new entries to a file using `since`, then I think the user should have to explicitly state the file to which the entries should be appended. 

I think the `log` and `trace` method names are fine.

**Comment by ddebrunner** [comment]
`since` wouldn't append to an existing file, instead it would be something like:

`retrieve_log_trace(since="11:00")` - Results in a file named `Job_18_1100.tgz`
`retrieve_log_trace(since="12:00")` - Results in a file named `Job_18_1200.tgz`

(with `11:00`, `12:00` being shorthand in this example for a full date).
Real details tbd.

**Comment by ddebrunner** [comment]
So the long name that the console uses is actually generated by the JMX API.

**Comment by ejpring** [comment]
Oh, come on! These are methods for Python programmers we're talking about. Everyone who uses these methods will be familiar with filename arguments. If they want the file stored in another directory, they don't need another argument, they'll just prepend a path to the filename. If they want a unique name for each file, they'll append a counter or a timestamp to the filename. 

I think its important for the default filename to be simple and predictable, something that does not surprise the programmer and send him or her scrambling to the documentation. I really can't imagine a situation where a programmer would be pleased to find randomly generated filenames stored in their file system. 

If you insist that the default names must be unique, then please insist that they be made unique with a timestamp so they can be sorted.

**Comment by ejpring** [comment]
Incremental log retrieval won't be as simple as adding a 'since' parameter to the methods we're discussing. The 'startTime' value in an HTTP GET request must be relative to the server's system clock, not the client's clock: the client needs to receive the 'Last-Modified-Milliseconds' header from the previous HTTP GET response, cache it, and use it in the next HTTP GET request. 

Incremental log retrieval needs a different set of methods that set up persistent HTTP connections and return live streams to their callers, rather than storing files. That probably requires changes on the server side as well. I think that's beyond the scope of this discussion.

**Comment by ejpring** [comment]
In any event, if we are worried that someone might use characters in a job name that are unacceptable in a file name, and we want default filenames to be unique per call, and we want them to be sortable, then we need something like this:

**Job::retrieve_logs_traces(filename=None)**
Retrieve console logs and application traces from all of the job's PEs and store them as a compressed 'tar' archive in the specified file. If not specified, the default filename is 'job_id_timestamp.tar.gz' in the current directory, where 'id' is the identifying number assigned to the job when it was submitted and 'timestamp' is the Unix epoch time when the file was stored (see https://en.wikipedia.org/wiki/Unix_time), according to the client's system clock. The method returns the name of the file it stored.

**PE::retrieve_console_log(filename=None)**
Retrieve the PE's console log and store it as uncompressed text in the specified file. If not specified, the default filename is 'pe_id_timestamp.log' in the current directory, where 'id' is the identifying number assigned to the PE when the scheduler started it and 'timestamp' is the Unix epoch time when the file was stored (see https://en.wikipedia.org/wiki/Unix_time), according to the client's system clock. The method returns the name of the file it stored.

**PE::retrieve_application_trace(filename=None)**
Retrieve the PE's application trace and store it as uncompressed text in the specified file. If not specified, the default filename is 'pe_id_timestamp.trace' in the current directory, where 'id' is the identifying number assigned to the PE when the scheduler started it and 'timestamp' is the Unix epoch time when the file was stored (see https://en.wikipedia.org/wiki/Unix_time), according to the client's system clock.. The method returns the name of the file it stored.

So, for example, these methods might store their files with names like these:

job_7_1511274857.tar.gz
pe_11_1511274859.log
pe_11_1511274862.trace

**Comment by ddebrunner** [comment]
> if we are worried that someone might use characters in a job name that are unacceptable in a file name
Just to note the default job names do exactly this, so it's not a rare situation.

**Comment by ejpring** [comment]
Are you thinking of the colon character in default job names like 'TestApplication::Main_7'? Colon characters are valid in filenames, and files with names like 'TestApplication::Main_0.tar.gz' work fine with GUI file management utilities. The Linux 'tar' command, however, mis-interprets the colon as indicating the file is remote, and tries to connect to a server. If that's a problem, you can work around it with the '--force-local' option of the 'tar' command, or use the alternative command 'bsdtar3' instead.

**Comment by wmarshall484** [comment]
> If they want the file stored in another directory, they don't need another argument, they'll just prepend a path to the filename.

I don't found this to be true. The `tempfile` module [has a few methods](https://docs.python.org/2/library/tempfile.html#tempfile.TemporaryFile) which take a `dir` parameter, independent from the name of the file. I find this to be analogous to our problem, where we want to store the logs somewhere, but we don't care about the name (but we want it to be unique). I find the following to be pythonic:

Job::retrieve_logs_traces(filename=None, dir = None)
Retrieve console logs and application traces from all of the job's PEs and store them as a compressed 'tar' archive `with the specified filename in the provided directory`. If not specified, the default filename is 'job_id_timestamp.tar.gz', where 'id' is the identifying number assigned to the job when it was submitted and 'timestamp' is the Unix epoch time when the file was stored (see https://en.wikipedia.org/wiki/Unix_time), according to the client's system clock. `If not specified, the logs are saved to the current directory.` The method returns the `full path` of the file it stored.

PE::retrieve_console_log(filename=None, dir = None)
Retrieve the PE's console log and store it as uncompressed text `with the specified filename in the provided directory`. If not specified, the default filename is 'pe_id_timestamp.log', where 'id' is the identifying number assigned to the PE when the scheduler started it and 'timestamp' is the Unix epoch time when the file was stored (see https://en.wikipedia.org/wiki/Unix_time), according to the client's system clock. `If not specified, the logs are saved to the current directory.` The method returns the `full path` of the file it stored.

PE::retrieve_application_trace(filename=None, dir = None)
Retrieve the PE's application trace and store it as uncompressed text `with the specified filename in the provided directory`. If not specified, the default filename is 'pe_id_timestamp.trace' in the current directory, where 'id' is the identifying number assigned to the PE when the scheduler started it and 'timestamp' is the Unix epoch time when the file was stored (see https://en.wikipedia.org/wiki/Unix_time), according to the client's system clock. `If not specified, the logs are saved to the current directory.` The method returns the `full path` of the file it stored.

**Comment by ejpring** [comment]
TemporaryFile() strikes me as a pretty obscure precedent. However, I can see that you have your heart set on it, so I have added a second argument 'dir' to all three methods. The first argument 'path' works like the familiar precedent 'open()'.

The development build/submit/logs/cancel lifecycle is complicated enough to need a sample application, so I have added one. Its similar to the unit tests without their clutter.
